### PR TITLE
Disallow Imperfect Collapse and Nested Parallelism with Data Dependencies

### PR DESCRIPTION
**Pull request by Atrisan**
## Description
Additional repetitions of write accesses can generate flaky or wrong behaviour on later parallel code, by destroying the conditions of maps.
The fix disallows an imperfect collapse where an explicit dependency between structured control flow elements is also an argument to the loop nest.

A similar issue for nested offloading regarding the repetition of accumulations is also fixed.

## Related Issues

- Closes #

## Type of Change
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Refactor / internal change (no functional change)
- [ ] Documentation update

## Quality Checklist
*Please ensure the following are completed before requesting review:*

* [x] Runtime-compatible (no externally visible existing method has its parameters changed, was deleted or has its exception behavior changed)
* [x] Compile-time compatible (new arguments & properties added have default values, but users need to recompile against the new version to work)

IR:
* [ ] Introduces new types into the SDFG IR
* [ ]  Changed / added properties on SDFG IR Elements (LibraryNodes, ControlFlowNodes DataFlowNodes etc.)
  *  [ ] New / modified types have serialization / deserialization support
  *  [ ] Deserialization is backward compatible (new serialization can deserialize old format and represent it in the new format if applicable)
  *  [ ] Serialized format is backward compatible (previous deserializer can read new serialized format without incorrectness (just lacking additional guarantees that are optional or did not exist before)
  *  [ ] Serializer has an option to emit the previous format

### Unit Tests
- [x] New unit tests have been added for the changes.
- [x] Existing unit tests pass locally.
- [x] Edge cases and error paths are covered.

### Integration Tests
- [x] New integration tests have been added (or existing ones updated).
- [x] Integration tests pass locally.
- [x] Interactions with other components/services are verified.

### Documentation
- [ ] Public APIs, options, and behavior changes are documented.
- [x] README / docs site / inline comments updated where applicable.
- [ ] Changelog / release notes updated (if applicable).

### Test Coverage
- [x] Test coverage has not decreased as a result of this change.
- [x] Coverage report reviewed and acceptable.
- [x] Critical / new code paths are covered.

## How Has This Been Tested?
Extended unit tests and the integration tests for MLP and Segformer.

## Additional Context
Add any other context about the PR here (design decisions, trade-offs, follow-ups).

### CODE REVIEW

File: opt/include/sdfg/transformations/map_collapse.h [hunk 1]
```diff
@@ -48,9 +48,15 @@ class MapCollapse : public Transformation {
...
-    bool check_imperfect(analysis::AnalysisManager& analysis_manager);
+    /// body element (its writes vary across the inner index), a write-write
+    /// conflict between two different body elements on the same container, or a
+    /// skipped element that performs a read-modify-write (e.g. a reduction
+    /// accumulator) on a container that escapes the loop. A read-modify-write on a
+    /// local (a container whose lifetime is confined to the loop, as reported by
+    /// ArgumentsAnalysis) is privatized per thread and does not violate parallel
+    /// access, so it - like a plain store, even to a function argument - remains
+    /// allowed.
+    bool check_imperfect(builder::StructuredSDFGBuilder& builder, analysis::AnalysisManager& analysis_manager);
```
Comment (NoraHagmeyer) [inline opt/include/sdfg/transformations/map_collapse.h]: I don't think you need the builder here. You definitely shouldn't need to for a check.

File: opt/tests/transformations/map_collapse_test.cpp [hunk 2]
```diff
@@ -2182,6 +2184,477 @@ TEST(MapCollapseTest, CannotApply_Imperfect_WriteWriteConflict) {
...
 
+// ---------------------------------------------------------------------------
+// Imperfect — replication safety (loop-local write model)
+// ---------------------------------------------------------------------------
+
+TEST(MapCollapseTest, CannotApply_Imperfect_SkippedReductionRecomputed) {
+    // Softmax-style hazard. A skipped sequential reduction accumulates sum[i] (a
+    // read-modify-write into a shared transient buffer), and the collapsible inner
+    // map divides by sum[i]. The buffer escapes the loop (it is consumed by a node
+    // after the collapsed region, exactly like the softmax normaliser _tmp_17), so
```
Comment (NoraHagmeyer) [inline opt/tests/transformations/map_collapse_test.cpp]: This is highly specific for the one sdfg you found the problem in. Unnecessary in the general code comment.

**Review by NoraHagmeyer — CHANGES_REQUESTED** [review]
Just 2 small things.

**Review by NoraHagmeyer — APPROVED** [review]
